attrs_strict/_type_validation.py

```diff
@@ -78,6 +80,70 @@ def _validate_elements(attribute, value, expected_type):
         _handle_tuple(attribute, value, expected_type)
     elif base_type == typing.Union:
         _handle_union(attribute, value, expected_type)
+    elif base_type in SimilarTypes.Callable:
+        _handle_callable(attribute, value, expected_type)
+
+
+def _get_base_type(type_):
+    if hasattr(type_, "__origin__") and type_.__origin__ is not None:
+        base_type = type_.__origin__
+    elif is_newtype(type_):
+        base_type = type_.__supertype__
+    else:
+        base_type = type_
+
+    return base_type
+
+
+def _type_matching(actual, expected):
+    actual = _get_base_type(actual) if is_newtype(actual) else actual
+    expected = _get_base_type(expected) if is_newtype(expected) else expected
+
+    if expected == actual or expected == typing.Any:
+        return True
+
+    base_type = _get_base_type(expected)
+
+    if base_type == typing.Union:
+        if any(
+            _type_matching(actual, expected_candidate)
+            for expected_candidate in expected.__args__
+        ):
+            return True
+
+    elif base_type in (
+        SimilarTypes.Dict
+        | SimilarTypes.List
+        | SimilarTypes.Tuple
+        | SimilarTypes.Callable
+    ):
+        if all(
+            _type_matching(actual, expected)
+            for actual, expected in zip_longest(
+                actual.__args__, expected.__args__
+            )
+        ):
+            return True
+
+    return False
+
+
+def _handle_callable(attribute, callable_, expected_type):
+    _signature = signature(callable_)
+    callable_args = [
+        param.annotation for param in _signature.parameters.values()
+    ]
+    callable_args.append(_signature.return_annotation)
+    if not expected_type.__args__:
+        return  # No annotations specified on type, matches all Callables
+
+    for callable_arg, expected_arg in zip_longest(
+        callable_args, expected_type.__args__
+    ):
+        if not _type_matching(callable_arg, expected_arg):
+            raise CallableError(
```

Comment:
If you think moving this into _type_matching is a good idea, you could use ```https://docs.python.org/2/library/functools.html#functools.partial``` to rewrite the signature into something simpler (as you'd have to pass in all those arguments all the time and that could be annoying).